File: snake-strategy-bono/src/main/java/co/electric/snake/strategy/bonostrategy/BlockingDirectionContainer.java
package co.electric.snake.strategy.bonostrategy;

import co.electric.snake.framework.model.Coordinate;
import co.electric.snake.framework.model.Direction;

import java.util.*;

public class BlockingDirectionContainer {

    private final Map<Direction, Integer> distanceToDirection = new HashMap<>();
    private final Map<Direction, Coordinate> coordinateToDirection = new HashMap<>();
    private final Map<Coordinate, Integer> distanceToCoordinate = new HashMap<>();
    private final Map<Coordinate, Direction> directionToCoordinate = new HashMap<>();

    public void putData(Direction direction, Coordinate coordinate, int distance) {
        if (!directionHasDistance(direction)) {
            doPut(direction, coordinate, distance);
        } else {
            final int storedDistanceToDirection = getDistanceByDirection(direction);
            if (distance < storedDistanceToDirection) {
                doPut(direction, coordinate, distance);
            }
        }
    }

    public Set<Direction> getDirections() {
        return distanceToDirection.keySet();
    }

    public SimpleDirectionContainer getFarthestBlockingDirections() {
        final Map<Integer, SimpleDirectionContainer> directionToDistances = new TreeMap<>(Collections.reverseOrder());
        distanceToDirection.forEach((direction, distance) -> {
            final SimpleDirectionContainer simpleDirections = Optional.ofNullable(directionToDistances.get(distance)).orElse(new SimpleDirectionContainer());
            simpleDirections.add(direction);
            directionToDistances.put(distance, simpleDirections);
        });
        return directionToDistances.values().stream()
                .findFirst()
                .orElse(new SimpleDirectionContainer());
    }

    private void doPut(Direction direction, Coordinate coordinate, int distance) {
        if (!coordinateHasDirection(coordinate)) {
            putDistanceToDirection(direction, distance);
            putCoordinateToDirection(direction, coordinate);
            putDistanceToCoordinate(coordinate, distance);
            putDirectionToCoordinate(coordinate, direction);
        } else {
            final int storedDistanceToCoordinate = getDistanceByCoordinate(coordinate);
            if (distance <= storedDistanceToCoordinate) {
                removeDataByCoordinate(coordinate);
                putDistanceToDirection(direction, distance);
                putCoordinateToDirection(direction, coordinate);
                putDistanceToCoordinate(coordinate, distance);
                putDirectionToCoordinate(coordinate, direction);
            }
        }
    }

    private void removeDataByCoordinate(Coordinate coordinate) {
        final Direction directionToCoordinate = getDirectionByCoordinate(coordinate);
        removeFromDistanceToDirection(directionToCoordinate);
        removeFromCoordinateToDirection(directionToCoordinate);
        removeFromDistanceToCoordinate(coordinate);
        removeFromDirectionToCoordinate(coordinate);
    }

    private void removeFromDistanceToDirection(Direction direction) {
        distanceToDirection.remove(direction);
    }

    private void removeFromCoordinateToDirection(Direction direction) {
        coordinateToDirection.remove(direction);
    }

    private void removeFromDistanceToCoordinate(Coordinate coordinate) {
        distanceToCoordinate.remove(coordinate);
    }

    private void removeFromDirectionToCoordinate(Coordinate coordinate) {
        directionToCoordinate.remove(coordinate);
    }

    private void putDistanceToDirection(Direction direction, int distance) {
        distanceToDirection.put(direction, distance);
    }

    private void putCoordinateToDirection(Direction direction, Coordinate coordinate) {
        coordinateToDirection.put(direction, coordinate);
    }

    private void putDistanceToCoordinate(Coordinate coordinate, int distance) {
        distanceToCoordinate.put(coordinate, distance);
    }

    private void putDirectionToCoordinate(Coordinate coordinate, Direction direction) {
        directionToCoordinate.put(coordinate, direction);
    }

    private int getDistanceByDirection(Direction direction) {
        return distanceToDirection.get(direction);
    }

    private int getDistanceByCoordinate(Coordinate coordinate) {
        return distanceToCoordinate.get(coordinate);
    }

    private Direction getDirectionByCoordinate(Coordinate coordinate) {
        return directionToCoordinate.get(coordinate);
    }

    private boolean directionHasDistance(Direction direction) {
        return distanceToDirection.containsKey(direction);
    }

    private boolean coordinateHasDirection(Coordinate coordinate) {
        return coordinateToDirection.containsValue(coordinate);
    }

    @Override
    public String toString() {
        return "DirectionData [distanceToDirection=" + distanceToDirection + "]";
    }

}

File: snake-strategy-bono/src/main/kotlin/co/electric/snake/strategy/bonostrategy/BlockingDirectionContainer.kt
package co.electric.snake.strategy.bonostrategy

import co.electric.snake.framework.model.Coordinate
import co.electric.snake.framework.model.Direction
import java.util.*
import kotlin.collections.HashMap

class BlockingDirectionContainer {

    private val distanceToDirection: MutableMap<Direction, Int> = EnumMap(Direction::class.java)
    private val coordinateToDirection: MutableMap<Direction, Coordinate> = EnumMap(Direction::class.java)
    private val distanceToCoordinate: MutableMap<Coordinate, Int> = HashMap()
    private val directionToCoordinate: MutableMap<Coordinate, Direction> = HashMap()

    fun putData(direction: Direction, coordinate: Coordinate, distance: Int) {
        if (!directionHasDistance(direction)) {
            doPutData(direction, coordinate, distance)
        } else {
            val storedDistanceToDirection = getDistanceByDirection(direction)
            if (distance < storedDistanceToDirection) {
                doPutData(direction, coordinate, distance)
            }
        }
    }

    fun getDirections(): Set<Direction> {
        return distanceToDirection.keys
    }

    fun getFarthestBlockingDirections(): SimpleDirectionContainer {
        val directionToDistances: MutableMap<Int, SimpleDirectionContainer> = TreeMap(Collections.reverseOrder())
        distanceToDirection.forEach { (direction: Direction, distance: Int) ->
            val simpleDirections = Optional.ofNullable(directionToDistances[distance]).orElse(SimpleDirectionContainer())
            simpleDirections.add(direction)
            directionToDistances[distance] = simpleDirections
        }
        return directionToDistances.values.stream()
                .findFirst()
                .orElse(SimpleDirectionContainer())
    }

    private fun doPutData(direction: Direction, coordinate: Coordinate, distance: Int) {
        if (!coordinateHasDirection(coordinate)) {
            putDistanceToDirection(direction, distance)
            putCoordinateToDirection(direction, coordinate)
            putDistanceToCoordinate(coordinate, distance)
            putDirectionToCoordinate(coordinate, direction)
        } else {
            val storedDistanceToCoordinate = getDistanceByCoordinate(coordinate)
            if (distance <= storedDistanceToCoordinate) {
                removeDataByCoordinate(coordinate)
                putDistanceToDirection(direction, distance)
                putCoordinateToDirection(direction, coordinate)
                putDistanceToCoordinate(coordinate, distance)
                putDirectionToCoordinate(coordinate, direction)
            }
        }
    }

    private fun directionHasDistance(direction: Direction): Boolean {
        return distanceToDirection.containsKey(direction)
    }

    private fun coordinateHasDirection(coordinate: Coordinate): Boolean {
        return coordinateToDirection.containsValue(coordinate)
    }

    private fun getDistanceByDirection(direction: Direction): Int {
        return Optional.ofNullable(distanceToDirection[direction]).orElse(Int.MAX_VALUE)
    }

    private fun getDistanceByCoordinate(coordinate: Coordinate): Int {
        return Optional.ofNullable(distanceToCoordinate[coordinate]).orElse(Int.MAX_VALUE)
    }

    private fun getDirectionByCoordinate(coordinate: Coordinate): Direction? {
        return directionToCoordinate[coordinate]
    }

    private fun putDistanceToDirection(direction: Direction, distance: Int) {
        distanceToDirection[direction] = distance
    }

    private fun putCoordinateToDirection(direction: Direction, coordinate: Coordinate) {
        coordinateToDirection[direction] = coordinate
    }

    private fun putDistanceToCoordinate(coordinate: Coordinate, distance: Int) {
        distanceToCoordinate[coordinate] = distance
    }

    private fun putDirectionToCoordinate(coordinate: Coordinate, direction: Direction) {
        directionToCoordinate[coordinate] = direction
    }

    private fun removeDataByCoordinate(coordinate: Coordinate) {
        val directionToCoordinate = getDirectionByCoordinate(coordinate)
        removeFromDistanceToDirection(directionToCoordinate)
        removeFromCoordinateToDirection(directionToCoordinate)
        removeFromDistanceToCoordinate(coordinate)
        removeFromDirectionToCoordinate(coordinate)
    }

    private fun removeFromDistanceToDirection(direction: Direction?) {
        distanceToDirection.remove(direction)
    }

    private fun removeFromCoordinateToDirection(direction: Direction?) {
        coordinateToDirection.remove(direction)
    }

    private fun removeFromDistanceToCoordinate(coordinate: Coordinate) {
        distanceToCoordinate.remove(coordinate)
    }

    private fun removeFromDirectionToCoordinate(coordinate: Coordinate) {
        directionToCoordinate.remove(coordinate)
    }

    override fun toString(): String {
        return "DirectionData [distanceToDirection=$distanceToDirection]"
    }

}
